src/influx_line.erl

```diff
@@ -0,0 +1,626 @@
+%% @end
+-spec encode(
+  Name::string()|atom()|binary(),
+  Points::list(influx_data_point())|influx_data_point(),
+  Tags::influx_data_point()|binary(),
+  Time::non_neg_integer()|atom()
+) -> Line::binary() | {error, Reason::atom()}.
+encode(Name, #{} = Map, Tags, Time) ->
+  encode_item(Name, Map, Tags, Time);
+
+encode(Name, [{_, _}|_] = List, Tags, Time) ->
+  encode_item(Name, List, Tags, Time);
+
+encode(Name, [[{_, _}|_]|_] = Points, Tags, Time) ->
+  encode_list(Name, Points, Tags, Time);
+
+encode(Name, [#{}|_] = Points, Tags, Time) ->
+  encode_list(Name, Points, Tags, Time);
+
+encode(_, _, _, _) -> {error, invalid_data}.
+
+%% @doc
+%% Returns current UTC time in nanoseconds
+%% @end
+-spec timestamp() -> Time::non_neg_integer().
+timestamp() ->
+  ulitos:timestamp()*1000000.
+
+%% ------------------------------------------------------------------
+%% Internal Function Definitions
+%% ------------------------------------------------------------------
+encode_item(Name, Item, Tags, Time) ->
+  BTags = encode_tags(Tags),
+  Measurement = to_key(Name),
+  Prefix = <<Measurement/binary, BTags/binary>>,
+  encode_item_with_prefix(Prefix, Item, Time).
+
+encode_item_with_prefix(Prefix, Item, Time) ->
+  BTime = encode_time(Time),
+  Fields = encode_fields(Item),
+  concat_line(Prefix, Fields, BTime).  
+
+encode_list(Name, [Item], Tags, Time) ->
+  encode_item(Name, Item, Tags, Time);
+
+encode_list(Name, List, Tags, Time) when is_integer(Time) ->
+  BTags = encode_tags(Tags),
+  Measurement = to_key(Name),
+  Prefix = <<Measurement/binary, BTags/binary>>,
+  encode_list_with_time(undefined, List, Prefix, Time, []);
+
+encode_list(Name, List, Tags, _Time) ->
+  encode_list(Name, List, Tags, influx_line:timestamp()).
+
+encode_list_with_time({error, _} = Error, _, _, _, _) ->
+  Error;
+
+encode_list_with_time(Item, [], _, _, Acc) ->
+  lists:foldl(fun concat_lines/2, ?EMPTY, [Item|Acc]);
+
+encode_list_with_time(undefined, [H|Rest], Prefix, Time, Acc) ->
+  encode_list_with_time(
+    encode_item_with_prefix(Prefix, H, Time),
+    Rest,
+    Prefix,
+    Time + 1,
+    Acc
+  );
+
+encode_list_with_time(Item, [H|Rest], Prefix, Time, Acc) ->
+  encode_list_with_time(
+    encode_item_with_prefix(Prefix, H, Time),
+    Rest,
+    Prefix,
+    Time + 1,
+    [Item|Acc]
+  ).
+
+encode_points_list({error, _} = Error, _, _, _) ->
+  Error;
+
+encode_points_list(Item, [], _, Acc) ->
+  lists:foldl(fun concat_lines/2, ?EMPTY, [Item|Acc]);
+
+encode_points_list(undefined, [H|Rest], Time, Acc) ->
+  encode_points_list(
+    encode_with_time(H, Time),
+    Rest,
+    Time + 1,
+    Acc
+  );
+
+encode_points_list(Item, [H|Rest], Time, Acc) ->
+  encode_points_list(
+    encode_with_time(H, Time),
+    Rest,
+    Time + 1,
+    [Item|Acc]
+  ).
+
+encode_with_time([{_,_}|_] = List, BaseTime) ->
+  M = proplists:get_value(measurement, List, undefined),
+  F = proplists:get_value(fields, List, undefined),
+  if M =:= undefined orelse F =:= undefined
```

Comment:
Replace the 'if' expression on line 219 with a 'case' expression or function clauses.